### ISSUE TITLE
403 Response on Login

### ISSUE THREAD

**Issue by Spacefreak18** [bbsan2k/plugin.video.f1tv#61]
Starting this weekend i am noticing a 403 error on login.

It appears the AppleTV app is having the same issue.

[https://github.com/NoahFetz/F1AppleTV/issues/55](url)

It appears to be on the server side.

**Comment by pabohoney1** [comment]
Seeing the same thing.  Would be great to get a patch in to watch Melbourne this weekend!  Thanks!

**Comment by Gghuzr** [comment]
It seems F1TV blocked 3rd party apps from login - official F1TV app is working on the iPhone

**Comment by RealMurphy** [comment]
Yeah, other apps are reporting not only you now need a session key to log-in which at least looks like do-able, but they also activated DRM which means you need to have a player capable of decrypting it (i.e. ffmpeg, vlc, ... would not directly work).

(see f1viewer/racecontrol projects for more details)

**Comment by Spacefreak18** [comment]
I just did more research and I'm seeing the same thing. I played around with the apiKey etc, in the headers with no success.

But like you say, even if i were to get a login, which I think I could, playing the actual video if it is DRM locked is going to be a different story.

> Yeah, other apps are reporting not only you now need a session key to log-in which at least looks like do-able, but they also activated DRM which means you need to have a player capable of decrypting it (i.e. ffmpeg, vlc, ... would not directly work).
> 
> (see f1viewer/racecontrol projects for more details)

**Comment by bbsan2k** [comment]
Okay, tbh I was expecting this to happen already quite some time.

Regarding the DRM: there is a good chance that this can be somehow done by using inputstream and widevine etc, but for now I‘m unfortunately not planning to look into it any further because of other project and also family time.

If anyone is interested in digging deeper into fixing stuff I‘d be happy to assist with understanding the current code or merging PRs if required.

**Comment by Spacefreak18** [comment]
> Okay, tbh I was expecting this to happen already quite some time.
> 
> Regarding the DRM: there is a good chance that this can be somehow done by using inputstream and widevine etc, but for now I‘m unfortunately not planning to look into it any further because of other project and also family time.
> 
> If anyone is interested in digging deeper into fixing stuff I‘d be happy to assist with understanding the current code or merging PRs if required.

no worries, i  know it's not anyone's fault. And I appreciate the work you've already done. :)

If you can post something so i can do some research on inputstream and widevine and how it handles DRM, that would be helpful for myself or someone else who might want to take a crack (pun intended) at implementing it.

**Comment by joost-quro** [comment]
This is really sad from F1 TV, because there a still another of devices that are not supported.  If you are able  to fix it that would be really great.

**Comment by acleee** [comment]
It looks like the DRM streams use VMP (Verified Media Path) - so I _think_ playback could only work on Android, since VMP requires the executable to be signed by Widevine's key (or something along those lines).

Anywho, this .strm file works and I was able to watch qualifying live on Android (account-specific tokens redacted; tested on my Note20 w/ Kodi 19):

```
#KODIPROP:inputstream=inputstream.adaptive
#KODIPROP:inputstream.adaptive.manifest_type=mpd
#KODIPROP:inputstream.adaptive.license_type=com.widevine.alpha
#KODIPROP:inputstream.adaptive.license_key=https://f1tv.formula1.com/2.0/R/ENG/WEB_HLS/ALL/CONTENT/LA/widevine?channelId=1033&contentId=1000005121|ascendontoken=xxxxxxx&entitlementtoken=xxxxxxx|A{SSM}|
https://ott-video-cf.formula1.com/out/v1/index.mpd?kid=1042&exp=1649571380&ttl=1440&token=xxxxxxxxx&start=1649483704```
```

I don't know if you'd want other headers on the license request to make it appear fully legit, but it's only checking the ascendontoken and entitlementtoken. These params can in theory be scripted/set within the addon (this is how the other WV kodi addons work)

The same strm file returns this error on the key request in Windows:

```
{"resultCode":"OK","message":"Downstream Licence Acquisition Failure","errorDescription":"ACN_5002","resultObj":{"keyos":{"keyosHttpStatus":403,"errorcode":371000005,"errormsg":"Widevine license acquisition failed >> Verified media path cannot be verified"}
```

**Comment by Gghuzr** [comment]
Who can step in at this project, Now? @TheDevFreak maybe?

**Comment by TheDevFreak** [comment]
> Who can step in at this project, Now? @TheDevFreak maybe?

Since F1TV now wants DRM I see little point in patching up the Kodi plugin, frankly since they've now released Android & Apple TV apps that work fine this plugin seems like it is replaced by those really...

**Comment by Gghuzr** [comment]
@TheDevFreak 

Unfortunately not for Linux/Libreelec and Enigma2 Users..

I would very appreciate for any further Support. This Plugin is working like a charm on the unsupported devices before

**Comment by TheDevFreak** [comment]
> @TheDevFreak
> 
> Unfortunately not for Linux/Libreelec and Enigma2 Users..
> 
> I would very appreciate for any further Support. This Plugin is working like a charm on the unsupported devices before



As I don't use the addon anymore and it has been so long since I worked on it I honestly don't remember the internals of it let alone how to get widevine to work fully in Kodi.

I know there is another F1TV addon by dut-iptv and they have an open issue and looks like they're adding support here: https://github.com/dut-iptv/dut-iptv/issues/124

**Comment by dazzler22** [comment]
thanks for your efforts over the years guys, it has been much appreciated, I have moved to a laptop with hdmi and logging in on the web, it looks like the changes that f1tv have introduced may have killed off the kodi addon (outside of android devices) 

Another option if your on a Pi4 is to move to android tv!

**Comment by Gghuzr** [comment]
thanks for your efforts I will follow dut-iptv if they can handle the livestreams at least

**Comment by lenlennart** [comment]
> Another option if your on a Pi4 is to move to android tv!

Unfortunately, there is no reliable AndroidTV for Pi4... At least non i am aware of :-( Konstakangs builds offer no hardware acceleration

Anyway, watching F1TV over laptop with Chromecast works as well, but the features in handling are not comparable to Kodi at all - please bring this addon back!

**Comment by archulet** [comment]
> Okay, tbh I was expecting this to happen already quite some time.
> 
> Regarding the DRM: there is a good chance that this can be somehow done by using inputstream and widevine etc, but for now I‘m unfortunately not planning to look into it any further because of other project and also family time.
> 
> If anyone is interested in digging deeper into fixing stuff I‘d be happy to assist with understanding the current code or merging PRs if required.
----------

@bbsan2k and @TheDevFreak please comment if necessary to cooperate trough paypal or other service, we understand the value of your time, also, you are the ones that understand the passion of F1 followers, because was the passion that move you start this plugin in the first place.  Either way many thanks for what we have had so far. Best regards!

**Comment by Spacefreak18** [comment]
for those looking for an interim solution, assuming you're not running kodi on some sort of android device, you can install web viewer add on which will open a browser in fullscreen kiosk mode.

On Linux if you aren't running kodi from a window manager, you'll have to play around to get it actually fullscreen. I don't know how i did it, i started a fluxbox session, setup chromium, installed window resizer extension and set it up, and then switched back to my kodi session.

https://iwf1.com/kodi/repository.sandmann79-py3.plugins

**Comment by Spacefreak18** [comment]
i actually tried to look at the headers sent by the android app on my tablet using mitmproxy, but it seems the app uses cert pinning which is becoming more and more common.
